Add support for NestedTensor share_memory_

adabeyta opened the pull request:
Fixes: https://github.com/pytorch/pytorch/issues/161915

### Summary

Implements share_memory_() support for NestedTensor! 

### Changes

- Added share_memory_() method to NestedTensor class.
  - Shares storage for all NestedTensor components: _values, _offsets, _lengths, and cached seqlen tensors.
  - Guard for CUDA Tensors.

### Testing

Before Fix:

`pytest -q test/test_nestedtensor.py -k "test_share_memory" -v`

Output:

```
Running 1 items in this shard

test/test_nestedtensor.py Fatal Python error: Segmentation fault
```

After Fix:

`pytest -q test/test_nestedtensor.py -k "test_share_memory" -v`

Output:

```
Running 1 items in this shard

test/test_nestedtensor.py::TestNestedTensorDeviceTypeCPU::test_share_memory_cpu PASSED [0.0753s] 
```

adabeyta committed:
Add support for NestedTensor share_memory

adabeyta committed:
Skip test_share_memory_cpu under TorchDynamo compilation

jbschlosser reviewed (commented):
Thanks for the contribution!

adabeyta committed:
Use DRY approach to call share_memory_()

jbschlosser commented on test/test_nestedtensor.py:
```diff
@@ -1230,6 +1230,25 @@ def test_device_checks(self, device):
+        nt = torch.nested.nested_tensor([a, b], layout=torch.jagged)
+
+        # Guard CUDA tensors
+        if device == "cuda":
```
```suggestion
        if "cuda" in device:
```

this is more accurate because in practice `device` will be something like `"cuda:0"`

jbschlosser commented on torch/nested/_internal/nested_tensor.py:
```diff
@@ -163,6 +163,19 @@ def offsets(self):
     def lengths(self):
         return self._lengths
 
+    def share_memory_(self):
```
instead of implementing the method directly on the subclass, it's better to override it via NJT's `__torch_function__()` implementation here: https://github.com/pytorch/pytorch/blob/f638854e1da6b33e78dcc9f3e28c98c4cdce4e86/torch/nested/_internal/ops.py#L370

this should compose better with the rest of PyTorch.

jbschlosser reviewed (commented):
couple small things, nothing huge :) thanks again!

jbschlosser commented:
> Does the NestedTensor.is_shared() method return the correct result? Or we must use NestedTensor.values().is_shared() to check?

We definitely want `NestedTensor.is_shared()` to return the correct result. @adabeyta this will likely require an additional implementation of `is_shared()` in NJT's `torch_function` impl.

adabeyta committed:
Override via NJT's torch_func impl

adabeyta commented:
> > Does the NestedTensor.is_shared() method return the correct result? Or we must use NestedTensor.values().is_shared() to check?
> 
> We definitely want `NestedTensor.is_shared()` to return the correct result. @adabeyta this will likely require an additional implementation of `is_shared()` in NJT's `torch_function` impl.

I went ahead and added NT.is_shared() to return the correct result and added it under the NJT impl as well. Thank you for catching that @Phimos

adabeyta committed:
Use NestedTensor.is_shared() instead of internal component checks

jbschlosser reviewed (approved):
Looks great, thanks for the updates!
